denonavr/ssdp.py

```diff
@@ -13,51 +13,51 @@
 import xml.etree.ElementTree as ET
 from urllib.parse import urlparse
 import requests
+import ifaddr
 
 _LOGGER = logging.getLogger('DenonSSDP')
 
 SSDP_ADDR = "239.255.255.250"
 SSDP_PORT = 1900
+SSDP_MX = 2
+SSDP_TARGET = (SSDP_ADDR, SSDP_PORT)
 SSDP_ST_1 = "ssdp:all"
 SSDP_ST_2 = "upnp:rootdevice"
 SSDP_ST_3 = "urn:schemas-upnp-org:device:MediaRenderer:1"
 
-SSDP_QUERY_1 = (
-    "M-SEARCH * HTTP/1.1\r\n" +
-    "HOST: {addr}:{port}\r\n".format(addr=SSDP_ADDR, port=SSDP_PORT) +
-    "MAN: \"ssdp:discover\"\r\n" +
-    "MX: 2\r\n" +
-    "ST: {st}\r\n".format(st=SSDP_ST_1) + "\r\n"
-)
-SSDP_QUERY_2 = (
-    "M-SEARCH * HTTP/1.1\r\n" +
-    "HOST: {addr}:{port}\r\n".format(addr=SSDP_ADDR, port=SSDP_PORT) +
-    "MAN: \"ssdp:discover\"\r\n" +
-    "MX: 2\r\n" +
-    "ST: {st}\r\n".format(st=SSDP_ST_2) + "\r\n"
-)
-SSDP_QUERY_3 = (
-    "M-SEARCH * HTTP/1.1\r\n" +
-    "HOST: {addr}:{port}\r\n".format(addr=SSDP_ADDR, port=SSDP_PORT) +
-    "MAN: \"ssdp:discover\"\r\n" +
-    "MX: 2\r\n" +
-    "ST: {st}\r\n".format(st=SSDP_ST_3) + "\r\n"
-)
-
-SSDP_QUERIES = (SSDP_QUERY_1, SSDP_QUERY_2, SSDP_QUERY_3)
+SSDP_ST_LIST = (SSDP_ST_1, SSDP_ST_2, SSDP_ST_3)
 
 SCPD_XMLNS = "{urn:schemas-upnp-org:device-1-0}"
 SCPD_DEVICE = "{xmlns}device".format(xmlns=SCPD_XMLNS)
 SCPD_DEVICETYPE = "{xmlns}deviceType".format(xmlns=SCPD_XMLNS)
 SCPD_MANUFACTURER = "{xmlns}manufacturer".format(xmlns=SCPD_XMLNS)
 SCPD_MODELNAME = "{xmlns}modelName".format(xmlns=SCPD_XMLNS)
+SCPD_SERIALNUMBER = "{xmlns}serialNumber".format(xmlns=SCPD_XMLNS)
 SCPD_FRIENDLYNAME = "{xmlns}friendlyName".format(xmlns=SCPD_XMLNS)
 SCPD_PRESENTATIONURL = "{xmlns}presentationURL".format(xmlns=SCPD_XMLNS)
 
 DEVICETYPE_DENON = "urn:schemas-upnp-org:device:MediaRenderer:1"
 
 SUPPORTED_MANUFACTURERS = ["Denon", "DENON", "Marantz"]
 
+def ssdp_request(ssdp_st, ssdp_mx=SSDP_MX):
+    """Return request bytes for given st and mx."""
+    return "\r\n".join([
+        'M-SEARCH * HTTP/1.1',
+        'ST: {}'.format(ssdp_st),
+        'MX: {:d}'.format(ssdp_mx),
+        'MAN: "ssdp:discover"',
+        'HOST: {}:{}'.format(*SSDP_TARGET),
+        '', '']).encode('utf-8')
+
+def get_local_ips():
+    adapters = ifaddr.get_adapters()
+    ips = []
+    for adapter in adapters:
+        for ip in adapter.ips:
+            if type(ip.ip)==str:
```

Comment:
`if isinstance(ip.ip, str)` please

Reply:
checking this is not neccessary anyhow. What else should it be?

Reply:
> checking this is not neccessary anyhow. What else should it be?

adapter.ips includes both Ipv4 and Ipv6 addresses, the Ipv6 addresses are tuples instead of strings and the code can't deal with Ipv6, so I filter those out.